shared/markdown/parser.pegjs

```diff
@@ -1,154 +1,161 @@
+EscapedChar
+ = EscapeMarker char:SpecialChar { return char }
 
-InsideItalicMarker
- = ((! ItalicMarker) .) { return text(); }
+NormalChar
+ = !NativeEmojiCharacter !SpecialChar NonBlank { return text() }
 
-InsideStrikeMarker
- = ((! StrikeMarker) .) { return text(); }
+Text
+ = NormalChar+ { return text() }
 
-InsideCodeBlock
- = ((! Ticks3) .) { return text(); }
+ // TODO: other unknown chars
 
-InsideInlineCode
- = ((! Ticks1) .) { return text(); }
-
-InsideQuoteBlock
- = ((! LineTerminatorSequence) .) { return text(); }
-
-// Here we use the literal ":" because we want to not match the :foo in ::foo
-InsideEmojiMarker
- = (! ":" [a-zA-Z0-9+_-]) { return text(); }
-
-InsideEmojiTone
- = "::skin-tone-" [1-6] { return text(); }
-
-// Define the rules for styles. Usually a start marker, children, and an end marker.
+// TODO: should coalesce multiple line quotes
 QuoteBlock
- = QuoteBlockMarker _? children:FromQuote* LineTerminatorSequence { return {type: 'quote-block', children}; }
+ = QuoteBlockMarker WhiteSpace* children:__INLINE_MACRO__<!LineTerminatorSequence> LineTerminatorSequence { return {type: 'quote-block', children} }
 
 Bold
- = BoldMarker children:FromBold* BoldMarker { return {type: 'bold', children}; }
+ = BoldMarker !WhiteSpace children:__INLINE_MACRO__<!BoldMarker> BoldMarker !(BoldMarker / NormalChar) { return {type: 'bold', children: flatten(children)} }
 
 Italic
- = ItalicMarker children:FromItalic* ItalicMarker { return {type: 'italic', children}; }
+ = ItalicMarker !WhiteSpace children:__INLINE_MACRO__<!ItalicMarker> ItalicMarker !(ItalicMarker / NormalChar) { return {type: 'italic', children: flatten(children)} }
 
 Strike
- = StrikeMarker children:FromStrike* StrikeMarker { return {type: 'strike', children}; }
+ = StrikeMarker !WhiteSpace children:__INLINE_MACRO__<!StrikeMarker> StrikeMarker !(StrikeMarker / NormalChar) { return {type: 'strike', children: flatten(children)} }
 
 CodeBlock
- = Ticks3 children:InsideCodeBlock* Ticks3 { return {type: 'code-block', children}; }
+ = Ticks3 children:(!Ticks3 .)+ Ticks3 { return {type: 'code-block', children: flatten(children)} }
 
 InlineCode
- = Ticks1 children:InsideInlineCode* Ticks1 { return {type: 'inline-code', children}; }
+ = Ticks1 children:(!Ticks1 .)+ Ticks1 { return {type: 'inline-code', children: flatten(children)} }
 
-Emoji
- = EmojiMarker children:InsideEmojiMarker+ tone:InsideEmojiTone? ":" { return {type: 'emoji', children: [children.join('') + (tone || '')]}; }
+// Here we use the literal ":" because we want to not match the :foo in ::foo
+InsideEmojiMarker
+ = !EmojiMarker [a-zA-Z0-9+_-] { return text() }
 
-Text "text"
- = _? NonBlank+ _? {
-    return convertLink(text())
- }
+InsideEmojiTone
+ = "::skin-tone-" [1-6] { return text() }
 
-// Useful helpers
+Emoji
+ = EmojiMarker children:InsideEmojiMarker+ tone:InsideEmojiTone? EmojiMarker { return {type: 'emoji', children: [text()]} }
+
+NativeEmojiCharacter "unicode emoji"
+ = [__EMOJI_CHARACTERS__]
```

Comment:
Note: this results in pretty big regex and array literals generated in the parser source. It doesn't seem easy to hoist these, but doing so could have a positive impact on perf. I didn't pursue it very far because hoisting the other regexes got us back to roughly the same perf as the previous parser.